lib/ui/text.dart

```diff
@@ -1489,6 +1490,24 @@ class ParagraphBuilder extends NativeFieldWrapperClass2 {
   }
   String _addText(String text) native 'ParagraphBuilder_addText';
```

Comment:
docs for `addText` should say what happens if `text` contains a U+FFFC character.

Reply:
assuming that having U+FFFC characters here will mess up where inline boxes end up, we should probably replace occurrences of U+FFFC in `text` with U+FFFD. Actually checking that seems like it would be an annoying cost to pay all the time though... maybe we can just set a flag when we call addBox (or whatever we call it), and only do the substitution during the buildParagraph() when we know we have called addBox, and otherwise do nothing?

Reply:
The way it is currently implemented, the U+FFFC characters are appended to the end of the text buffer, making it ambiguous which U+FFFC was inserted through addText() and which through addBox() at build-time. I could store the indexes separately to indicate which instances are boxes and which should be replaced at build-time

Reply:
if passing U+FFFC through addText will break things, and we have no way to prevent it, we have to document the heck out of this. That's a serious problem. It'll mean that someone who uses addPlaceholder at all and happens to put user-generated content in there too, they'll be broken by any user who (intentionally or not) includes a U+FFFC in their text.

I think we need to find an efficient way to not break people though.

Reply:
This should be ok now as we no longer are completely dependent on the existence U+FFFC, the full indicator for a placeholder run is "single codepoint run with U+FFFC, and the index needs to be in the known placeholders list"